ligo/main.mligo

```diff
@@ -207,21 +225,118 @@ let set_position (s : storage) (i_l : tick_index) (i_u : tick_index) (i_l_l : ti
+    // Binary search, invariant: buffer[l].time <= t && t < buffer[r].time
+    if l_i + 1n < r_i
+    then
+        let m_i = (l_i + r_i) / 2n in
+        let m_v = get_registered_cumulatives_unsafe buffer m_i in
+        if m_v.time > t
+        then find_cumulatives_around (buffer, t, l_i, m_i, l_v, m_v)
+        else find_cumulatives_around (buffer, t, m_i, r_i, m_v, r_v)
+    else
+        (l_v, r_v, assert_nat (t - l_v.time, internal_observe_bin_search_failed))
+
+let get_cumulatives (buffer : timed_cumulatives_buffer) (t : timestamp) : cumulatives_value =
+    let l_i = buffer.first in
+    let r_i = buffer.last in
+    let l_v = get_registered_cumulatives_unsafe buffer l_i in
+    let r_v = get_registered_cumulatives_unsafe buffer r_i in
+
+    let _: unit = if t < l_v.time then failwith observe_outdated_timestamp_err else unit in
+    let _: unit = if t > r_v.time then failwith observe_future_timestamp_err else unit in
+    let (sums_at_left, sums_at_right, time_delta) = find_cumulatives_around (buffer, t, l_i, r_i, l_v, r_v)
+
+    // When no updates to contract are performed, time-weighted accumulators grow
+    // linearly. Extrapolating to get the value at timestamp in-between.
+    //
+    // tick_cumulative(t) and liqudity_per_second_cumulative(t) functions produced
+    // by this extrapolation are continuous.
+    // 1. At [left, right) range found by the binary search above, cumulatives are
+    //    continuous by construction - our extrapolation is linear.
+    // 2. At (right - o, right] range they are also continous, because we will
+    //    use the same formula for calculating cumulatives at `right - o` (here)
+    //    and at `right` (see how `sum` fields are updated in `update_timed_cumulatives`).
+    in  { tick_cumulative =
+            let at_left_block_end_tick_value = sums_at_right.tick.block_start_value
+            in sums_at_left.tick.sum + time_delta * at_left_block_end_tick_value.i
+        ; seconds_per_liquidity_cumulative =
+            let at_left_block_end_lps_value = sums_at_right.lps.block_start_liquidity_value
+            in {x128 = sums_at_left.lps.sum.x128 +
+                Bitwise.shift_left time_delta 128n / at_left_block_end_lps_value }
+        }
+
+let observe (s : storage) (p : observe_param) : result =
+    let value = List.map (get_cumulatives s.cumulatives_buffer) p.times
+    in ([Tezos.transaction value 0mutez p.callback], s)
+
+// Update the cumulative values stored for the recent timestamps.
+//
+// This has to be called on every update to the contract, not necessarily
+// for each block. Currently all cumulatives keep time-weighted sum of something,
+// so we can extrapolate these cumulative on periods of the contract's inactivity.
+let update_timed_cumulatives (s : storage) : storage =
+    let buffer = s.cumulatives_buffer in
+
+    let last_value = get_last_cumulatives buffer in
+    (* Update not more often than once per block *)
```

Comment:
Is this logic actually correct? If the contract receives multiple updates in the same block shouldn't the the stored data here be representative of the state after the last update?

E.g. if the `curr_tick_index` gets updated more than once shouldn't its last value be the one that contributes to the sum?

Reply:
I seem to account for this. E.g. below in this function, `cur_tick_index` is recorded into `tick.block_start_value`, and treated in the other code respectively.

Like, value after the last update in the block is the same as the value before any updates in the next block.

Reply:
Hm, indeed there is probably another option - update those values every time (that will be a step away from what the initial implementation did). If you think it may be simpler and clearer, I can think over it and write down the comparison.

Reply:
Oh, I see, I was actually confused about this. 

Following the original whitepaper, the value for a block track the state _before_ said block, so any change following it doesn't matter and should only affect the following block(s).

Got it, in this case the current implementation makes sense to me and I think we should keep it :+1: